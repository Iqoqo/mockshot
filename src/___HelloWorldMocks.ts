export class HelloWorldMocks {
    static bar;

    static bar(mock: "no-valid-params" | "success"): any {
        switch (mock) {
            case "no-valid-params":
                return {
                  "error": "yes"
                }
            case "success":
                return {
                  "api": {
                    "data": "123"
                  }
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }

    static shape;

    static shape(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "data": {
                    "_id": "1234"
                  },
                  "id": "1234sddd56",
                  "some": "value"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }

    static foo;

    static foo(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "foo": "barr"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }
}
