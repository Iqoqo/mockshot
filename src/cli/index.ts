import { ApiGenerator } from "../generators/ApiGenerator";
import { GeneratorRunner } from "../GeneratorRunner";
import { SnapshotFetcher } from "./util";

export async function run(args: string[]) {
  const outputDir = args[0];

  const snapshots = SnapshotFetcher.getSnapshots();

  const generators = [new ApiGenerator()];

  const generatorRunner = new GeneratorRunner(outputDir, generators, snapshots);
  await generatorRunner.run();
}
