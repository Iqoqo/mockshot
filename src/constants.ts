export const MockshotTag = "[mockshot]";
export const Success = "success";
