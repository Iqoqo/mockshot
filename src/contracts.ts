export interface ISnapshot {
  key: string;
  packageName?: string;
  filePath: string;
  data: any;
}

export interface IApiSnapshot extends ISnapshot {
  data: IApiSnapData;
}

export interface IClassSnapshot extends ISnapshot {
  data: IClassSnapData;
}

export interface IApiSnapData extends IApiSnapDataBase {
  mockName: string;
}

export interface IClassSnapData {
  className: string;
  methodName: string;
  mockName: string;
  mock: any;
}

export interface IApiSnapDataBase {
  url: string;
  httpMethod: "post" | "get" | "put" | "delete" | "patch";
  mock: {
    statusCode: number;
    body: any;
    error?: string;
  };
}

export type MatcherReturn = {
  pass: boolean;
  message(): string | (() => string);
};
