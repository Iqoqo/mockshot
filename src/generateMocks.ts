import {glob} from "glob";
import * as path from "path";
import Project from "ts-simple-ast";
import fs from "fs";
import { SnapshotState } from "jest-snapshot";
import pretty from 'json-pretty'

export async function generateMocks() {
  const project = new Project({
    compilerOptions: { outDir: "dist/mocks", declaration: true }
  });

  glob("**/*.snap", async function(er, files) {
    // files is an array of filenames.
    // If the `nonull` option is set, and nothing
    // was found, then files is ["**/*.js"]
    // er is an error object or null.
    const mockDef = {} as any;

    files.forEach(file => {
      const absolutePath = path.resolve(file);
      const state = new SnapshotState(absolutePath, {
        snapshotPath: absolutePath
      });
      const data = state._snapshotData;
      const keys = Object.keys(data);

      keys.forEach(async key => {
        if (key.indexOf("[mockshot]") < 0) {
          return;
        }
        const snapshot = JSON.parse(state._snapshotData[key]);
        let classDef = mockDef[snapshot.className];
        if (!classDef) {
          classDef = {};
          mockDef[snapshot.className] = classDef;
        }
        let methodDef = classDef[snapshot.methodName];
        if (!methodDef) {
          methodDef = {};
          classDef[snapshot.methodName] = methodDef;
        }

        let mockName = methodDef[snapshot.mockName];
        if (!mockName) {
          methodDef[snapshot.mockName] = snapshot.mock;
        } else {
          throw Error(
            "Duplicate mock name for method " +
              snapshot.methodName +
              ": " +
              snapshot.mockName
          );
        }
      });
    });

    const classNames = Object.keys(mockDef);
    classNames.forEach(async className => {
      const mockClassName = className+'Mocks';

      console.log('Creating class', mockClassName);

      const mockFileName = `mocks/${mockClassName}.ts`;

      let myClassFile;
      try {
        await fs.unlinkSync(mockFileName);
        console.log("File exists", mockFileName, "removing...");
      } catch (ex) {
      }

      myClassFile = project.createSourceFile(mockFileName);

      const classDeclaration = myClassFile.addClass({
        name: mockClassName
      });

      classDeclaration.setIsExported(true);

      const methods = mockDef[className];

      const methodNames = Object.keys(methods);

      methodNames.forEach(methodName => {
        console.log('Creating method mocks', methodName);

        const mocks = methods[methodName];
        const mockNames = Object.keys(mocks);
        let mockTypes = mockNames.map(value => {
          return `"${value}"`;
        });
        const types = mockTypes.join(" | ");

        const method = classDeclaration.addMethod({
          isStatic: true,
          parameters: [{ name: "mock", type: types }],
          name: methodName,
          returnType: "any"
        });

        method.setBodyText(writer =>
          writer.write("switch (mock)").block(() => {
            mockNames.forEach(mockName => {
              writer.write(`case "${mockName}":`).indentBlock(()=>{
                writer.write(`return ${pretty(mocks[mockName])}`);
              })
            });
            writer.write(`default:`).indentBlock(()=>{
              writer.write(`throw Error("Unknown mock: "+mock);`);
            })
          })
        );
      });
    });

    await project.save();

    console.log('Saved Typescript source...emitting vanilla js');
    await project.emit();
    console.log('Done!');
  });
}
