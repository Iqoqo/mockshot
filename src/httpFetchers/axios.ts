class AxiosResponse extends BaseHttpFetcher {
  isValid(): boolean {
    return true;
  }
}
interface Axios {
  response: {
    config: {
      method: string;
      url: string;
    };
    status: number;
    data: any;
  };
}

function instanceOfA(fetchLibrary: any): fetchLibrary is Axios {
  return (
    "response" in object &&
    "response.config" in object &&
    response.config.method
  );
}
