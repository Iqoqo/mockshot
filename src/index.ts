export * from "./matchers";
export * from "./generators";
export { run } from "./run";
