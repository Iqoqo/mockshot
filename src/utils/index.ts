export function getSafePackageName(packageName: string): string {
  if (packageName.startsWith("@")) {
    return packageName.slice(1);
  }
  return packageName;
}
