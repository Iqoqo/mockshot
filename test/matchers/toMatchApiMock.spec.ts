import axios from "axios";
import adapter from "axios/lib/adapters/http";
import chai from "chai";
import chaiHttp from "chai-http";
import http from "http";
import "../../src";
import { getOwnSnapshots } from "../utils";

const port = "8123";
const testUrl = `http://localhost:${port}`;
const returnedObject = { hello: "world", foo: "bar" };
const urlQuery = "?abc=def";

const app = new http.Server();

app.on("request", (req, res) => {
  res.writeHead(200, { "Content-Type": "application/json" });
  res.write(JSON.stringify(returnedObject));
  res.end();
});

const getSnapshot = key =>
  getOwnSnapshots(__filename).find(snap => snap.key.includes(key));

describe("toMatchApiMock()", () => {
  beforeAll(() => {
    app.listen(port, () => {
      console.log(`listening on port ${port}`);
    });
  });
  afterAll(() => {
    app.close();
  });

  it("Should work with axios module ~1~", async () => {
    const path = "/users/123";
    const res = await axios.get(testUrl + path + urlQuery, { adapter });

    expect(res).toMatchApiMock("axios-success");
    const { data } = getSnapshot("~1~");

    expect(data.mock.body).toEqual(returnedObject);
    expect(data.mock.statusCode).toBe(200);
    expect(data.httpMethod).toBe("get");
    expect(data.mockName).toBe("axios-success");
    expect(data.url).toBe(path);
  });

  it("Should work with chai module ~2~", async () => {
    const path = "/login";
    chai.use(chaiHttp);
    const res = await chai
      .request(testUrl)
      .post(path + urlQuery)
      .send();

    expect(res).toMatchApiMock("chai-success");
    const { data } = getSnapshot("~2~");

    expect(data.mock.body).toEqual(returnedObject);
    expect(data.mock.statusCode).toBe(200);
    expect(data.httpMethod).toBe("post");
    expect(data.mockName).toBe("chai-success");
    expect(data.url).toBe(path);
  });

  it("Should use success mockName if none provided ~3~", async () => {
    const path = "/login";
    chai.use(chaiHttp);
    const res = await chai
      .request(testUrl)
      .post(path + urlQuery)
      .send();

    expect(res).toMatchApiMock();
    const { data } = getSnapshot("~3~");

    expect(data.mockName).toBe("success");
  });
});
