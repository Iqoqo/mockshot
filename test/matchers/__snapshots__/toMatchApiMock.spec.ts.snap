// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchApiMock() Should use success mockName if none provided ~3~: [mockshot] [APISnap] [post /login success] 1`] = `
{
  "httpMethod": "post",
  "mock": {
    "body": {
      "foo": "bar",
      "hello": "world"
    },
    "statusCode": 200
  },
  "mockName": "success",
  "url": "/login"
}
`;

exports[`toMatchApiMock() Should work with axios module ~1~: [mockshot] [APISnap] [get /users/123 axios-success] 1`] = `
{
  "httpMethod": "get",
  "mock": {
    "body": {
      "foo": "bar",
      "hello": "world"
    },
    "statusCode": 200
  },
  "mockName": "axios-success",
  "url": "/users/123"
}
`;

exports[`toMatchApiMock() Should work with chai module ~2~: [mockshot] [APISnap] [post /login chai-success] 1`] = `
{
  "httpMethod": "post",
  "mock": {
    "body": {
      "foo": "bar",
      "hello": "world"
    },
    "statusCode": 200
  },
  "mockName": "chai-success",
  "url": "/login"
}
`;
